vllm/v1/engine/parallel_sampling.py

```diff
@@ -0,0 +1,370 @@
+        async for out in child_gen:
+            if req_out := self.process_output(out, index):
+                yield req_out
+
+    @property
+    def num_completions(self) -> int:
+        assert self.request_output is not None
+        return len(self.request_output.outputs)
+
+    @property
+    def n(self) -> int:
+        return self.sampling_params.n
+
+    @property
+    def output_kind(self) -> RequestOutputKind:
+        return self.sampling_params.output_kind
+
+
+class SyncParallelSamplingManager:
+
+    def __init__(self):
+        # Parent req ID -> parent request manager
+        self.parent_reqs: Dict[str, ParallelSamplingRequest] = {}
+        # Child req ID -> (child req index, parent req ID)
+        self.child_reqs: Dict[str, Tuple[int, str]] = {}
+        # Flag to reset parallel sampling bookkeeping logic
+        # between engine runs
+        self._do_reset = False
+
+    def _reset_if_needed(self) -> None:
+        """Reset at beginning of sync generate()"""
+        if self._do_reset:
+            self.parent_reqs.clear()
+            self.child_reqs.clear()
+            self._do_reset = False
+
+    def schedule_reset(self) -> None:
+        """Schedule reset for the next time a parent request is added."""
+        if self.parent_reqs:
+            self._do_reset = True
+
+    def register_parent_request(self, req: ParallelSamplingRequest) -> None:
+        """Register parallel sampling parent request."""
+        self._reset_if_needed()
+        self.parent_reqs[req.request_id] = req
+
+    def register_child_request(self, req_id: str, child_req_id: str,
+                               index: int) -> None:
+        """Register parallel sampling child request with parent.
+        
+        Args:
+          req_id: parent request ID
+          child_req_id: child request ID
+          index: child request index within `n` child requests
+        """
+        self.child_reqs[child_req_id] = (index, req_id)
+
+    def step(
+        self,
+        outputs: List[RequestOutput],
+    ) -> List[RequestOutput]:
+        """Build parallel sampling request outputs.
+        
+        Extract child request outputs, aggregate them
+        into parent request output, and return parent
+        output when complete.
+
+        Do not modify `n=1` requests.
+
+        Args:
+          outputs: step request outputs. Mix of child request
+                   outputs & `n=1` request outputs.
+
+        Return:
+          List of parallel sampling parent request outputs &
+          unmodified `n=1` request outputs passed-thru from input.
+        """
+        if not (self.parent_reqs and outputs):
+            # Return unmodified
+            return outputs
+        agg_outputs = []
+        for c_out in outputs:
+            c_req_id = c_out.request_id
+            if cdx_req_id := self.child_reqs.get(c_req_id, None):
+                # For each parallel sampling child request output:
+                (cdx, req_id) = cdx_req_id
```

Comment:
Use clearer var names?
```suggestion
        for output in outputs:
            req_id = output.request_id
            if child_req_entry := self.child_reqs.get(req_id, None):
                # For each parallel sampling child request output:
                index, parent_req_id = child_req_entry
```
...

Reply:
Agreed, fixed manually.